### ISSUE TITLE
CustomEventTracker not firing?

### ISSUE THREAD

**Issue by MrMilo** [imgly/imgly-sdk-android-demo#23]
I try to implement the CustomEventTracker but seems like there's no data sending to my Google Analytics.. 

I follow the tutorial at https://developers.google.com/analytics/devguides/collection/android/v4/ to add in the library in gradle.

and in the CustomEventTracker I added `private Tracker googleAnalyticsTracker;` below `private String trackerId;`

Sample of implementation:
```
public void onClick(View v) {
                SettingsList settingsList = new SettingsList();

                settingsList.setEventTracker(new CustomEventTracker("UA-12345678-1")); //I changed my GA code

                settingsList.getConfig().setTools(
                        new OrientationEditorTool(R.string.imgly_tool_name_orientation, R.drawable.imgly_icon_tool_orientation),
                        new Divider(),
                        new StickerEditorTool(R.string.imgly_tool_name_sticker, R.drawable.imgly_icon_tool_sticker)
                );

                settingsList
                        .getSettingsModel(CameraSettings.class)
                        .setExportDir(Directory.DCIM, FOLDER)
                        .setExportPrefix("camera_")
                        .getSettingsModel(EditorSaveSettings.class)
                        .setExportDir(Directory.DCIM, FOLDER)
                        .setExportPrefix("result_")
                        .setSavePolicy(
                                EditorSaveSettings.SavePolicy.RETURN_ALWAYS_ONLY_OUTPUT
                        );

                new CameraPreviewBuilder(getActivity())
                        .setSettingsList(settingsList)
                        .startActivityForResult(getActivity(), CAMERA_PREVIEW_RESULT);
            }
```

**Comment by MrMilo** [comment]
If I initialize `settingsList.setEventTracker(new CustomEventTracker("UA-12345678-1"))` at Application.java, GA is finally firing up, but still not tracking Event by the SDK, any clue? @maltebaumann

**Comment by MrMilo** [comment]
Hi I notice when I initialize the EventTracker, in `SettingsList.java` , `StateObservable settings = settingsList.get(eventTracker.getClass());` will be null, am I doing it wrong?

**Comment by maltebaumann** [comment]
Hi @MrMilo,
could you provide an example for one of the callback methods within your CustomEventTracker class? Your call to `setEventTracker()` seems fine to me.

**Comment by MrMilo** [comment]
Hi @maltebaumann 

In my CustomEventTracker.java I only added `private Tracker googleAnalyticsTracker;` and uncommitted everything. Should I add any other code to make the tracker work as expected?

**Comment by MrMilo** [comment]
Below is my CustomEventTracker.java

```
package com.magicmirror.aicodetechnology.magicmirror2;

import android.os.Parcel;

import com.google.android.gms.analytics.GoogleAnalytics;
import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;

import java.util.List;

import ly.img.android.ImgLySdk;
import ly.img.android.sdk.models.config.AbstractConfig;
import ly.img.android.sdk.models.state.BrushLayerSettings;
import ly.img.android.sdk.models.state.BrushLayerSettingsEvent;
import ly.img.android.sdk.models.state.ColorAdjustmentSettings;
import ly.img.android.sdk.models.state.ColorAdjustmentSettingsEvent;
import ly.img.android.sdk.models.state.CropSettings;
import ly.img.android.sdk.models.state.CropSettingsEvent;
import ly.img.android.sdk.models.state.EditorMenuState;
import ly.img.android.sdk.models.state.EditorMenuStateEvent;
import ly.img.android.sdk.models.state.FilterSettings;
import ly.img.android.sdk.models.state.FilterSettingsEvent;
import ly.img.android.sdk.models.state.FocusSettings;
import ly.img.android.sdk.models.state.FocusSettingsEvent;
import ly.img.android.sdk.models.state.LayerListSettings;
import ly.img.android.sdk.models.state.LayerListSettingsEvent;
import ly.img.android.sdk.models.state.OrientationSettings;
import ly.img.android.sdk.models.state.OrientationSettingsEvent;
import ly.img.android.sdk.models.state.StickerLayerSettings;
import ly.img.android.sdk.models.state.StickerLayerSettingsEvent;
import ly.img.android.sdk.models.state.manager.EventTracker;
import ly.img.android.sdk.models.state.manager.OnStateEvent;

@SuppressWarnings("unused")
public class CustomEventTracker extends EventTracker {

    private String trackerId;
    private Tracker googleAnalyticsTracker;

    public CustomEventTracker(String trackerId) {
        this.trackerId = trackerId;

        GoogleAnalytics analytics = GoogleAnalytics.getInstance(ImgLySdk.getAppContext());
        googleAnalyticsTracker = analytics.newTracker(trackerId);

    }

    @OnStateEvent(model = EditorMenuState.class, event = EditorMenuStateEvent.TOOL_STACK_CHANGED, onInvalidation = false, onRevert = false)
    protected void changeToolView(EditorMenuState menuState) {

        googleAnalyticsTracker.setScreenName(menuState.getCurrentTool().getName());
        googleAnalyticsTracker.send(new HitBuilders.ScreenViewBuilder().build());

    }

    @OnStateEvent(model = EditorMenuState.class, event = EditorMenuStateEvent.CANCEL_AND_LEAVE, onInvalidation = false, onRevert = false)
    protected void onLeaveToolWithCancel(EditorMenuState menuState) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("menu action")
                .setLabel(menuState.getCurrentTool().getName())
                .setAction("cancel");
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = EditorMenuState.class, event = EditorMenuStateEvent.ACCEPT_AND_LEAVE, onInvalidation = false, onRevert = false)
    protected void onLeaveToolWithAccept(EditorMenuState menuState) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("menu action")
                .setLabel(menuState.getCurrentTool().getName())
                .setAction("accept");
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = ColorAdjustmentSettings.class, event = ColorAdjustmentSettingsEvent.CONTRAST, onInvalidation = false, onRevert = false)
    protected void onColorAdjustmentChangeContrast(ColorAdjustmentSettings colorAdjustmentSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("change color adjustment")
                .setLabel("contrast")
                .setAction("changed")
                .setValue(Math.round(colorAdjustmentSettings.getContrast() * 100));
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = ColorAdjustmentSettings.class, event = ColorAdjustmentSettingsEvent.BRIGHTNESS, onInvalidation = false, onRevert = false)
    protected void onColorAdjustmentChangeBrightness(ColorAdjustmentSettings colorAdjustmentSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("change color adjustment")
                .setLabel("brightness")
                .setAction("changed")
                .setValue(Math.round(colorAdjustmentSettings.getBrightness() * 100));
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = ColorAdjustmentSettings.class, event = ColorAdjustmentSettingsEvent.SATURATION, onInvalidation = false, onRevert = false)
    protected void onColorAdjustmentChangeSaturation(ColorAdjustmentSettings colorAdjustmentSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("change color adjustment")
                .setLabel("saturation")
                .setAction("changed")
                .setValue(Math.round(colorAdjustmentSettings.getSaturation() * 100));
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = BrushLayerSettings.class, event = BrushLayerSettingsEvent.COLOR, onInvalidation = false, onRevert = false)
    protected void onBrushColorChange(BrushLayerSettings brushLayerSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("change brush")
                .setLabel("color")
                .setAction("changed")
                .setValue(brushLayerSettings.getBrushColor());
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = BrushLayerSettings.class, event = BrushLayerSettingsEvent.HARDNESS, onInvalidation = false, onRevert = false)
    protected void onBrushHardnessChange(BrushLayerSettings brushLayerSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("change brush")
                .setLabel("hardness")
                .setAction("changed")
                .setValue(Math.round(brushLayerSettings.getBrushHardness() * 100));
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = BrushLayerSettings.class, event = BrushLayerSettingsEvent.SIZE, onInvalidation = false, onRevert = false)
    protected void onBrushSizeChange(BrushLayerSettings brushLayerSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("change brush")
                .setLabel("size")
                .setAction("changed")
                .setValue(Math.round(brushLayerSettings.getBrushSize()));
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = CropSettings.class, event = CropSettingsEvent.ASPECT, onInvalidation = false, onRevert = false)
    protected void onCropAspectChanged(CropSettings cropSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("crop changed")
                .setLabel("aspect")
                .setAction("changed aspect to: "+cropSettings.getAspect().getName());
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = FilterSettings.class, event = FilterSettingsEvent.FILTER, onInvalidation = false, onRevert = false)
    protected void onFilterChanged(FilterSettings filterSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("filter changed")
                .setLabel("filter: " + filterSettings.getFilter().getName())
                .setAction("set fitler");
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = FilterSettings.class, event = FilterSettingsEvent.INTENSITY, onInvalidation = false, onRevert = false)
    protected void onFilterIntensityChanged(FilterSettings filterSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("filter changed")
                .setLabel("filter: " + filterSettings.getFilter().getName())
                .setAction("changed intensity to: " + filterSettings.getIntensity());
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = FocusSettings.class, event = FocusSettingsEvent.MODE, onInvalidation = false, onRevert = false)
    protected void onFocusTypeChanged(FocusSettings focusSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("focus changed")
                .setLabel("focus: " + focusSettings.getFocusMode())
                .setAction("set type");
        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = FocusSettings.class, event = FocusSettingsEvent.INTENSITY, onInvalidation = false, onRevert = false)
    protected void onFocusIntensityChanged(FocusSettings focusSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("focus changed")
                .setLabel("focus: " + focusSettings.getFocusMode())
                .setAction("set intensity")
                .setValue(Math.round(focusSettings.getFocusIntensity() * 100));

        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = OrientationSettings.class, event = OrientationSettingsEvent.FLIP_VERTICAL, onInvalidation = false, onRevert = false)
    protected void onVerticalFlipChanged(OrientationSettings orientationSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("orientation changed")
                .setLabel("Is vertical flipped: " + orientationSettings.isVerticalFlipped())
                .setAction("change vertical flip");

        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = OrientationSettings.class, event = OrientationSettingsEvent.FLIP_HORIZONTAL, onInvalidation = false, onRevert = false)
    protected void onHorizontalFlipChanged(OrientationSettings orientationSettings) {

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("orientation changed")
                .setLabel("Is horizontal flipped: " + orientationSettings.isHorizontalFlipped())
                .setAction("change horizontal flip");

        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = OrientationSettings.class, event = OrientationSettingsEvent.ROTATION_CW, onInvalidation = false, onRevert = false)
    protected void onRotationCW(OrientationSettings orientationSettings) {
        int rotation = orientationSettings.getRotation();

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("orientation changed")
                .setLabel("Image Rotation: " + rotation)
                .setAction("rotate CW");

        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = OrientationSettings.class, event = OrientationSettingsEvent.ROTATION_CCW, onInvalidation = false, onRevert = false)
    protected void onRotationCCW(OrientationSettings orientationSettings) {
        int rotation = orientationSettings.getRotation();

        HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                .setCategory("orientation changed")
                .setLabel("Image Rotation: " + rotation)
                .setAction("rotate CCW");

        googleAnalyticsTracker.send(builder.build());

    }

    @OnStateEvent(model = LayerListSettings.class, event = LayerListSettingsEvent.ADD_LAYER, onInvalidation = false, onRevert = false)
    protected void onStickerAdd(LayerListSettings settings) {

        List<LayerListSettings.LayerSettings> list = settings.getLayerSettingsList();
        LayerListSettings.LayerSettings lastLayer = list.get(list.size() - 1);
        if (lastLayer instanceof StickerLayerSettings) {
            AbstractConfig.StickerConfigInterface stickerConfig = ((StickerLayerSettings) lastLayer).getStickerConfig();

            HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                    .setCategory(stickerConfig.getType() == AbstractConfig.StickerConfigInterface.STICKER_TYPE.TEXT ? "text": "sticker")
                    .setLabel(stickerConfig.getName())
                    .setAction("add sticker");
            googleAnalyticsTracker.send(builder.build());
        }
    }

    @OnStateEvent(model = LayerListSettings.class, event = LayerListSettingsEvent.REMOVE_LAYER, onInvalidation = false, onRevert = false)
    protected void onStickerRemove(LayerListSettings settings) {

        List<LayerListSettings.LayerSettings> list = settings.getLayerSettingsList();
        LayerListSettings.LayerSettings lastLayer = list.get(list.size() - 1);
        if (lastLayer instanceof StickerLayerSettings) {
            AbstractConfig.StickerConfigInterface stickerConfig = ((StickerLayerSettings) lastLayer).getStickerConfig();

            HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                    .setCategory(stickerConfig.getType() == AbstractConfig.StickerConfigInterface.STICKER_TYPE.TEXT ? "text": "sticker")
                    .setLabel(stickerConfig.getName())
                    .setAction("remove sticker");
            googleAnalyticsTracker.send(builder.build());
        }
    }

    @OnStateEvent(model = StickerLayerSettings.class, event = StickerLayerSettingsEvent.FLIP_HORIZONTAL, onInvalidation = false, onRevert = false)
    protected void onStickerFlipHorizontal(LayerListSettings settings) {

        LayerListSettings.LayerSettings lastLayer = settings.getSelected();
        if (lastLayer instanceof StickerLayerSettings) {
            AbstractConfig.StickerConfigInterface stickerConfig = ((StickerLayerSettings) lastLayer).getStickerConfig();
            HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                    .setCategory(stickerConfig.getType() == AbstractConfig.StickerConfigInterface.STICKER_TYPE.TEXT ? "text": "sticker")
                    .setLabel(stickerConfig.getName())
                    .setAction("horizontal flip");
            googleAnalyticsTracker.send(builder.build());
        }
    }

    @OnStateEvent(model = StickerLayerSettings.class, event = StickerLayerSettingsEvent.FLIP_VERTICAL, onInvalidation = false, onRevert = false)
    protected void onStickerFlipVertical(LayerListSettings settings) {

        LayerListSettings.LayerSettings lastLayer = settings.getSelected();
        if (lastLayer instanceof StickerLayerSettings) {
            AbstractConfig.StickerConfigInterface stickerConfig = ((StickerLayerSettings) lastLayer).getStickerConfig();
            HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                    .setCategory(stickerConfig.getType() == AbstractConfig.StickerConfigInterface.STICKER_TYPE.TEXT ? "text" : "sticker")
                    .setLabel(stickerConfig.getName())
                    .setAction("vertical flip");
            googleAnalyticsTracker.send(builder.build());
        }
    }

    @OnStateEvent(model = StickerLayerSettings.class, event = StickerLayerSettingsEvent.PLACEMENT_INVALID, onInvalidation = false, onRevert = false)
    protected void onStickerPositionChange(LayerListSettings settings) {
        LayerListSettings.LayerSettings lastLayer = settings.getSelected();
        if (lastLayer instanceof StickerLayerSettings) {
            AbstractConfig.StickerConfigInterface stickerConfig = ((StickerLayerSettings) lastLayer).getStickerConfig();
            HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                    .setCategory(stickerConfig.getType() == AbstractConfig.StickerConfigInterface.STICKER_TYPE.TEXT ? "text": "sticker")
                    .setLabel(stickerConfig.getName())
                    .setAction("position changed");
            googleAnalyticsTracker.send(builder.build());
        }
    }

    @OnStateEvent(model = LayerListSettings.class, event = LayerListSettingsEvent.BRING_TO_FRONT, onInvalidation = false, onRevert = false)
    protected void onStickerToFront(LayerListSettings settings) {
        LayerListSettings.LayerSettings lastLayer = settings.getSelected();
        if (lastLayer instanceof StickerLayerSettings) {
            AbstractConfig.StickerConfigInterface stickerConfig = ((StickerLayerSettings) lastLayer).getStickerConfig();
            HitBuilders.EventBuilder builder = new HitBuilders.EventBuilder()
                    .setCategory(stickerConfig.getType() == AbstractConfig.StickerConfigInterface.STICKER_TYPE.TEXT ? "text": "sticker")
                    .setLabel(stickerConfig.getName())
                    .setAction("bring to front");
            googleAnalyticsTracker.send(builder.build());
        }
    }

    protected CustomEventTracker(Parcel in) {
        this(in.readString());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(this.trackerId);
    }

    public static final Creator<CustomEventTracker> CREATOR = new Creator<CustomEventTracker>() {
        @Override
        public CustomEventTracker createFromParcel(Parcel source) {
            return new CustomEventTracker(source);
        }

        @Override
        public CustomEventTracker[] newArray(int size) {
            return new CustomEventTracker[size];
        }
    };
}
```

**Comment by MrMilo** [comment]
I also placed a break point at here:

`public SettingsList setEventTracker(EventTracker eventTracker) {
    StateObservable settings = settingsList.get(eventTracker.getClass());
    if (settings != null) { //settings is null, so not initialized
        settingsList.put(eventTracker.getClass(), eventTracker);
        eventTracker.onBind(this);
    }
    return this;
}`

and noticed the settings are null, why is that so?

**Comment by svendvd** [comment]
Yes, you are right!
We have released v2.0.13 your problem should be fixed now.

Please note the changes of the `CustomEventTracker.java` example implementation, we have fixed the parcel constructor!

**Comment by MrMilo** [comment]
@svendvd thanks ! I found another issues after the update, will create another issues thread. this problem had been solved!
